apps/emqx/src/emqx_topic.erl

```diff
@@ -210,17 +210,14 @@ join([]) ->
 join([W]) ->
     bin(W);
 join(Words) ->
-    {_, Bin} = lists:foldr(
-        fun
-            (W, {true, Tail}) ->
-                {false, <<W/binary, Tail/binary>>};
-            (W, {false, Tail}) ->
-                {false, <<W/binary, "/", Tail/binary>>}
-        end,
-        {true, <<>>},
-        [bin(W) || W <- Words]
-    ),
-    Bin.
+    do_join(lists:reverse(lists:map(fun bin/1, Words)), {true, <<>>}).
```

Comment:
there is no need for the `true` flag as `Words` can never be empty nor can it be single element list (because of the clause at line `208` and line `210`）
so we can just implement it like this:

```
join([First | Rest]) ->
    do_join(bin(First), Rest).

do_join(Topic, []) -> Topic;
do_join(Topic, [Word | Words]) -> do_join(<<Topic/binary, “/”, bin(Word)/binary>>, Words).
```

Reply:
with this impl, we can delete the clause at line 210